### PR TITLE
#201: add unit tests and solve 3 issues

### LINKED ISSUE #199: Fix incorrect print out
Please fix this:
```
[lbsStatistics] 
Final imbalance: 0.04376020743219233
```

### LINKED ISSUE #206: Ensure that object_index in ObjectCommunicator is always used (should NOT be optional)
In particular in `lbsVTStatisticsReader`:
```
rank_obj.set_communicator(ObjectCommunicator(r=received, s=sent))
```
the object index should be set (and is available).

### PR DESCRIPTION
**Pull request by marcinwrobel1986**
### This PR:
- added tox configuration for tests
- added CI config for Unit testing and Acceptance testing
- solves issues #198 and #199 
- updated readme
- change logger transferring objects from info to debug (solves #200 )
- fixed formatting in LBAF
- removed TODO with Ensure that object_index in ObjectCommunicator is always used (should NOT be optional) with fixed #206 

Closes #199 #198 #200 #206

### CODE REVIEW

File: src/Model/lbsObjectCommunicator.py [hunk 1]
```diff
@@ -47,14 +47,11 @@
...
-    """A class holding received and sent messages for an object
+    """ A class holding received and sent messages for an object
     """
 
     def __init__(self, r: dict = None, s: dict = None, i=None, logger: Logger = None):
-        if r is None:
-            r = {}
-        if s is None:
-            s = {}
+        # TODO: if i is needed? What for? i is used only when populate from samplers
```
Comment (ppebay) [inline src/Model/lbsObjectCommunicator.py:54]: Yes, this is needed for a sanity check in `summarize_unidirectional()`:
```
            if k.get_id() == self.object_index:
                self.lgr.error(f"object {self.object_index} cannot send communication to itself.")
                sys.exit(1)
```
What is however a problem is that is **not** used when we are populating from the VT reader. This is probably what should be addressed.

File: src/Model/lbsObjectCommunicator.py [hunk 2]
```diff
@@ -67,46 +64,17 @@ def __init__(self, r: dict = None, s: dict = None, i=None, logger: Logger = None
...
-            sys.exit(1)
-
         # Initialize list of volumes
         volumes = []
 
         # Iterate over one-way communications
         communications = self.sent if direction == "to" else self.received
         for k, v in communications.items():
             # Sanity check
+            # TODO: if self.object_index will be removed, then it needs to be removed as well
```
Comment (ppebay) [inline src/Model/lbsObjectCommunicator.py:77]: This sanity check **is** needed

File: src/Model/lbsObjectCommunicator.py [hunk 3]
```diff
@@ -115,21 +83,43 @@ def summarize_unidirectional(self, direction, print_indent=None):
             volumes.append(v)
 
             # Report current communication item if requested
+            # TODO: is this indentation really needed?
```
Comment (ppebay) [inline src/Model/lbsObjectCommunicator.py:86]: I do not remember having created it. Please decide what looks best in the output and chose the better option. Tx

**Review by ppebay — CHANGES_REQUESTED** [review]
Please see my responses. Tx.

File: src/Model/lbsMessage.py [hunk 4]
```diff
@@ -43,20 +43,20 @@
...
-    """A class representing information sent between ranks
+    """ A class representing information sent between ranks
     """
 
     def __init__(self, r, c):
         # Member variables passed by constructor
         self.round = r
         self.content = c
-
+    # TODO: Are these getters needed? Instance members are not protected. They could be reached with Message(1, 2).round
```
[the remark this answers is not in the document]
Comment (marcinwrobel1986) [reply src/Model/lbsMessage.py:53]: ok, then I will change `self.round = r`, to `self.__round = r` that is `private` in Python

**Comment by marcinwrobel1986** [comment]
@ppebay done, please check

File: src/Applications/LBAF.py [hunk 5]
```diff
@@ -403,8 +400,11 @@ def get_output_file_stem(params, n_ranks):
         objects.sort(key=lambda x: x.get("id"))
 
         # Execute rank order enumerator and fetch optimal arrangements
-        n_a, w_min_max, a_min_max = roe.compute_min_max_arrangements_work(
-            objects)
+        alpha = params.work_model.get('parameters').get('alpha')
```
[the remark this answers is not in the document]
Comment (marcinwrobel1986) [reply src/Applications/LBAF.py]: done

File: src/IO/lbsVTStatisticsReader.py [hunk 6]
```diff
@@ -57,25 +56,9 @@
...
-    entries specify the communication volume between objects in bytes.
-
-    Load profile collection and output is enabled in VT with the following flags:
-
-      mpirun -n 4 ./program --vt_lb_stats
-                            --vt_lb_stats_dir=my-stats-dir
-                            --vt_lb_stats_file=<base-name>
+    """ A class to read VT Object Map files. These json files could be compressed with Brotli.
+        Each file is named as <base-name>.<node>.json, where <node> spans the number of MPI ranks that VT is utilizing.
+        Schema of the compatible files is defined in <project-path>src/IO/schemaValidator.py
```
[the remark this answers is not in the document]
Comment (marcinwrobel1986) [reply src/IO/lbsVTStatisticsReader.py]: done

**Review by marcinwrobel1986 — COMMENTED** [review]
sorted

File: src/Model/lbsObjectCommunicator.py (hunk 3 above)
Comment (marcinwrobel1986) [reply src/Model/lbsObjectCommunicator.py]: sorted

File: src/Model/lbsObjectCommunicator.py (hunk 1 above)
Comment (ppebay) [inline src/Model/lbsObjectCommunicator.py:54]: This `# TODO` should be removed

File: src/Model/lbsObjectCommunicator.py [hunk 7]
```diff
@@ -67,46 +64,17 @@ def __init__(self, r: dict = None, s: dict = None, i=None, logger: Logger = None
...
-            sys.exit(1)
-
         # Initialize list of volumes
         volumes = []
 
         # Iterate over one-way communications
         communications = self.sent if direction == "to" else self.received
         for k, v in communications.items():
             # Sanity check
+            # TODO: if self.object_index will be removed, then it needs to be removed as well
```
Comment (ppebay) [inline src/Model/lbsObjectCommunicator.py:77]: Remove `# TODO` plz

File: src/Model/lbsPhase.py [hunk 8]
```diff
@@ -56,11 +56,10 @@
 
 
 class Phase:
-    """A class representing the state of collection of ranks with
-    objects at a given round
+    """ A class representing the state of collection of ranks with objects at a given round
     """
 
-    def __init__(self, t=0, logger: Logger = None, logging_level: str = 'info', file_suffix="vom"):
```
Comment (ppebay) [inline src/Model/lbsPhase.py:63]: Double quotes for `info`

**Review by ppebay — CHANGES_REQUESTED** [review]

File: src/Model/lbsPhase.py (hunk 8 above)
Comment (marcinwrobel1986) [reply src/Model/lbsPhase.py:63]: Philippe this line no longer exists :)

File: src/Model/lbsObjectCommunicator.py (hunk 1 above)
Comment (marcinwrobel1986) [reply src/Model/lbsObjectCommunicator.py]: done with #206 solved

File: src/Model/lbsObjectCommunicator.py (hunk 7 above)
Comment (marcinwrobel1986) [reply src/Model/lbsObjectCommunicator.py]: done with #206 solved

**Review by ppebay — APPROVED** [review]

**Pull request merged**